File: flex+bison/pretty.c
#include <stdio.h>
#include "pretty.h"

void prettyEXP(EXP *e)
{
	switch (e->kind) {
		case k_expressionKindIdentifier:
			printf("%s", e->val.identifier);
			break;
		case k_expressionKindIntLiteral:
			printf("%i", e->val.intLiteral);
			break;
		case k_expressionKindAddition:
			printf("(");
			prettyEXP(e->val.binary.lhs);
			printf("+");
			prettyEXP(e->val.binary.rhs);
			printf(")");
			break;
		case k_expressionKindSubtraction:
			printf("(");
			prettyEXP(e->val.binary.lhs);
			printf("-");
			prettyEXP(e->val.binary.rhs);
			printf(")");
			break;
		case k_expressionKindMultiplication:
			printf("(");
			prettyEXP(e->val.binary.lhs);
			printf("*");
			prettyEXP(e->val.binary.rhs);
			printf(")");
			break;
		case k_expressionKindDivision:
			printf("(");
			prettyEXP(e->val.binary.lhs);
			printf("/");
			prettyEXP(e->val.binary.rhs);
			printf(")");
			break;
	}
}

File: ocamllex+menhir/pretty.ml
open Tree

let rec pretty_exp (e: exp) =
  match e with
  | Identifier (id, _) -> print_string id
  | IntLiteral (v, _) -> print_string (string_of_int v)
  | Binary (lhs, rhs, op) -> pretty_binary_exp lhs rhs op
and pretty_binary_exp (lhs: exp) (rhs: exp) (op: operator) =
  match op with
  | Plus ->
    print_string "(";
    pretty_exp lhs;
    print_string " + ";
    pretty_exp rhs;
    print_string ")"
  | Minus ->
    print_string "(";
    pretty_exp lhs;
    print_string " - ";
    pretty_exp rhs;
    print_string ")"
  | Mult ->
    print_string "(";
    pretty_exp lhs;
    print_string " * ";
    pretty_exp rhs;
    print_string ")"
  | Div ->
    print_string "(";
    pretty_exp lhs;
    print_string " / ";
    pretty_exp rhs;
    print_string ")"
;;

let pretty_program (prog: program) =
  match prog with
  | Program e -> pretty_exp e; print_string "\n"
;;